File: gca3/68.cpp
#include <string>
#include <vector>

using namespace std;

/*
This problem is very annoying -> just need to become faster at solving LC prblems -> faster and thinking!!

*/
class Solution {
  public:
    vector<string> fullJustify(vector<string> &words, int maxWidth) {
        vector<string> result;
        int n = words.size();

        vector<string> rowV;
        int length = 0;
        for (int i = 0; i < n; i++) {
            if (rowV.size() + length + words[i].size() > maxWidth) {
                string row = "";
                int intervals = rowV.size() - 1;
                int spaces = maxWidth - length;
                if (intervals > 0) {
                    int int_space = spaces / intervals;
                    int extra_space = spaces % intervals;

                    for (int j = 0; j < rowV.size(); j++) {
                        row += rowV[j];
                        if (j != rowV.size() - 1) {
                            int space_to_add = int_space + (j < extra_space ? 1 : 0);
                            row += string(space_to_add, ' ');
                        }
                    }
                } else {
                    row += rowV[0];
                    row += string(spaces, ' ');
                }
                result.push_back(row);
                rowV.clear();
                length = 0;
            }

            rowV.push_back(words[i]);
            length += words[i].size();
        }

        string lastRow = "";
        for (int j = 0; j < rowV.size(); j++) {
            lastRow += rowV[j];
            if (j != rowV.size() - 1) {
                lastRow += ' ';
            }
        }

        lastRow += string(maxWidth - lastRow.size(), ' ');
        result.push_back(lastRow);
        return result;
    }
};

File: src/archive/gca3/68.py
from typing import List

class Solution:
    def fullJustify(self, words: List[str], maxWidth: int) -> List[str]:
        result = []
        
        line, length = [], 0
        i = 0
        while i < len(words):
            if length + len(line) + len(words[i]) > maxWidth:
                # need to insert spaces and add to the result vector
                extraSpace = maxWidth - length
                spaces = extraSpace // max(1, len(line) - 1)
                remainder = extraSpace % max(1, len(line) - 1)

                for j in range(max(1, len(line) - 1)):
                    line[j] += " " * spaces
                    if remainder:
                        line[j] += " "
                        remainder -= 1
                    
                result.append("".join(line))
                line, length = [], 0

            line.append(words[i])
            length+=len(words[i])
            i += 1
        
        # last line
        lastLine = " ".join(line)
        trailSpace = maxWidth - len(lastLine)
        lastLine += " " * trailSpace
        result.append(lastLine)
        return result
        